### ISSUE TITLE
aliases

### ISSUE THREAD

**Issue by rain-1** [rain-1/s#14]
A common use case of aliases is `alias ls='ls --color=auto'`. How would this be achieved in `s`?

One way might be to have a local bin folder in $PATH which chainloads the original ls. (try this out)

**Comment by edvb** [comment]
While I do like and appreciate the simplicity of this design, I do not know how it will work in practice, having multiple files of the same name in a user's `$PATH` becomes messy.

``` bash
s$ echo "ls --color=auto" | > ~/bin/ls
s$ echo $PATH
/home/me/bin:/usr/local/bin:/usr/bin:/bin
s$ whereis ls
ls: /bin/ls /home/me/bin/ls /usr/share/man/man1/ls.1.gz
s$ which ls
/bin/ls
```

Chooses `/bin/ls` over the alias `/home/me/bin/ls`, does not work even if `~/bin/ls` is at end not beginning of `$PATH`:

``` bash
s$ echo $PATH
/usr/local/bin:/usr/bin:/bin:/home/ed/bin
s$ whereis ls
ls: /bin/ls /home/me/bin/ls /usr/share/man/man1/ls.1.gz
s$ which ls
/bin/ls
```

This method would work really well for functions though since they are used in a similar style to files, while aliases seem more like substitutions. For example `func mkcd "mkdir $1 && cd $1"` would create a file in `~/bin` with `mkdir $1 && cd $1`

Perhaps aliases could be similar to how builtins work, but instead of a constant array of a builtin struct it can be a linked list contating two strings which gets added to when `alias` is called and removed when `unalias` is called.

**Comment by rain-1** [comment]
I think it works if you do `chmod +x` though.

**Comment by edvb** [comment]
Forgot about that thanks, but now I am getting the error `/bin/sh: error while loading shared libraries: libc.so.6: cannot open shared object file: Error 24` which appears to be the error code for too many open files.

But I still think that this replicates functions more then aliases, in this example you also have to include `$*` in the alias to list give arugemts since it is a file not a substitution, which is not how other languages implement aliases but how they do functions.

**Comment by rain-1** [comment]
ah, could this be an infinite loop!

**Comment by edvb** [comment]
Yes I believe `ls` is now calling itself inside the new file. This is another issue if we are going to use this method for aliases, which often do override other commands.

I have created a quick builtin for `func` which creates a file in `sbin` (set to `./sbin/` right now, should probably be in `~/.sbin/` or `~/.config/s/bin/`) with the name given and the command inserted into the new file.

``` bash
$ set PATH $PATH:sbin
$ func la 'ls -a "$@"'
$ la docs
.  ..  GRAMMAR  notes.txt  README  sh.txt  strings.md  TODO  tokens.txt
```

``` c
const char *sbin = "sbin/";

void
builtin_func(char **argv)
{
	FILE *f;
	char path[PATH_MAX];

	if (!argv[1] || !argv[2])
		reportret(,"%s: two arguments required", argv[0]);

	strcat(path, sbin);
	strcat(path, argv[1]);
	if (!(f = fopen(path, "w")))
		reportret(,"%s: %s: could not create file", argv[0], path);

	fputs(argv[2], f);
	fclose(f);
	if (chmod(path, S_IRWXU))
		report("%s: %s: could not chmod: %s", argv[0], path, strerror(errno));
}
```

Currently it seems the file is being run in `sh` not `s` since it does not insert a shebang. What is the correct shebang for `s`? Maybe set to `#!${PREFIX}/bin/s` using `PREFIX` from `config.mk`? Or maybe have `s` `source` it directly to solve this and so `sbin` does not need to be in `$PATH`.

`alias` would simply be a wrapper to this function which also appends `"$@"` to the end to simulate the `alias` functionality in other shells.

Note that in the current proposal `$@` will need to be added as a variable that expands to all the given arguments to a script (`$1`, `$2`, etc). Could this be done without breaking the one token to one token philosophy, I am not sure how else this crucial part in making shell scripts can be implemented other wise. Maybe is something similar to how `rc` does this by making all variables arrays? This could probability be opened as a new issue, or added to #8.

**Comment by rain-1** [comment]
Would it be possible to implement func as an `s` script? We should avoid creating new builtins when possible.

**Comment by rain-1** [comment]
Made an issue about $@ https://github.com/rain-1/s/issues/15

**Comment by rain-1** [comment]
What about `#!/usr/bin/env s`? Then we don't need to get root to install the thing, just add it to PATH?

by the way, I broke comments when I rewrote the tokenizer before so the shebang line was breaking but the latest commit repairs them.

**Comment by edvb** [comment]
> Would it be possible to implement func as an s script?

I would recommend against separating core components of `s` into dependencies as it would just fragment the code base, especially when they can be integrated as a builtin in under 20 lines of C. Separating it into a supporting file or s script would just over complicate things and add too many unnecessary lines of code. Writing supporting s scripts should be saved for higher level enhancements and wrappers not core features.

I also argue that the `>`, `<`, and `>>` supporting programs should also be integrated directly into `s` as builtins, especially since they are already in C. Right now unless you install `s` as a root user they can not be used since they are not in `$PATH`. This is a problem for a necessary component such as redirects. Putting them into `s` would allow us to clean up the code better and require less lines, thus simplify the code base. Having certain features in external programs and others not seems to only complicate the code, consolidating code seems like a better approach.

> What about `#!/usr/bin/env s`

Good idea thanks.

> I broke comments when I rewrote the tokenizer before so the shebang line was breaking but the latest commit repairs them

Thanks I was wondering why they were not working for me

**Comment by ng-0** [comment]
Couldn't this be done in an config file, for example `~/.s.rc` and you declare aliases in there?
Of course this adds the requirement to somehow parse the file, but it seems like the most portable solution to me.

**Comment by edvb** [comment]
@ng-0 Adding a rc file is defiantly a feature I want to implement, it should be as easily as adding a `interpreter_loop()` for the file; however, I do not see how this solves the problem of adding aliases since we still have to implement them for it to work in a config file.
